### ISSUE TITLE
ActionMailer deliver_later ovewrites message_id in Rails 4.2.3

### ISSUE THREAD

**Issue by tarr11** [rails/rails#21220]
I am not sure how to put a test together for this, but it appears that `deliver_later` is overwriting the message_id after delivery

``` 
msg = Notifier.welcome(User.first)
msg.message_id = "foo"
msg.deliver_now
# check message_id on delivered mail, as expected

msg.deliver_later
# check message_id on delivered mail (ie, check development.log), it has changed
```

I'd expect `deliver_later` to not change the message_id

**Comment by BenEddy** [comment]
This issue arises because the original `msg` instance is discarded. `deliver_later` serializes the mailer class, mailer method and arguments, and deserializes them later when the email is sent.

```Ruby
module ActionMailer
  class DeliveryJob < ActiveJob::Base # :nodoc:
    def perform(mailer, mail_method, delivery_method, *args) #:nodoc:
      # mailer is 'Notifier'
      # mail_method is 'welcome'
      # delivery_method is 'deliver_now'
      # args is the User

      mailer.constantize.public_send(mail_method, *args).send(delivery_method)
    end
  end
end
```

The expression `mailer.constantize.public_send(mail_method, *args)` effectively runs `Notifier.welcome(User.first)` again. It instantiates a new message, so setting the `message_id` on the original `msg` has no effect.

##### Fixing the issue

The easiest "fix" is to set the "Message-ID" header via the `headers` hash in the `welcome` method.

```Ruby
class Notifier
  def welcome(user, message_id)
    headers['message-id'] = message_id

    # stuff
  end
end
```

The harder fix requires framework modification. Rails would need to serialize the original `msg` when enqueuing the mail job.

`Mail::Message` has support for serialized to and from YAML. How offensive is the idea of serializing entire original message? It should be noted that changing the serialized representation of an delivery job would break preexisting enqueued jobs (unless special care was taken to ensure backward compatibility).

**Comment by BenEddy** [comment]
Also, maybe it's worth pointing out that this issue is a symptom of a potentially bigger issue: `deliver_later` results in running mailer methods twice. I could see this confusing some people. I don't think anyone would do this, but constructs like the following would be problematic:

```Ruby
class Notifier
  def welcome(user, message_id)
    user.increment!(:notification_count)
  end
end
```

Serializing the original `Mail::Message` ensure that the method is only execute once. The downside would be having large YAML string jammed into the job backend.

**Comment by PhilCoggins** [comment]
Just now running into this issue when setting the `smtp_envelope_to` instance variable on a `Message` object. This is problematic as you can't set smtp envelope fields in the header of an email message.

Are there any alternatives to serializing the message objects to YAML?

**Comment by BenEddy** [comment]
In the short term (and maybe you already did this), I think you can reference `@_message` in the body of your mailer method.

```Ruby
class Notifier
  def welcome(user, smtp_envelope_to)
    @_message.smtp_envelope_to = smtp_envelope_to

    # stuff
  end
end
```

This is definitely a terrible solution, though. Nothing screams framework-internals-coupling quite like referencing an underscore-prefixed ivar.

**Comment by mgodwin** [comment]
Since the `mail` method returns the message, you can actually do this without referencing the underscore-prefixed ivar:

```ruby
def mailer_method(smtp_envelope_to)
  message = mail(subject: 'blah')
  message.smtp_envelope_to = 'joe@example.com'
  message
end
```

I'd prefer to see this fixed a little bit better, has bitten us a few times now. 👍

**Comment by jeremy** [comment]
> deliver_later results in running mailer methods twice

`Notifier.welcome(*args)` returns a proxy object; it doesn't call `#welcome`. When you call `#deliver_later` on the proxy, it enqueues a job with the mailer and args. It hasn't run the mailer method at all at this point.

However, if you invoke methods on the proxy it will delegate them to the message, which lazy loads by calling the mailer method: `#welcome`. So, you'll get double-calls to the mailer method if you interact with the message then deliver it later.

Passing along the serialized message is attractive, but also fraught with issues.

What we could do… lock down the proxy object. Don't allow `#deliver_later` if you've lazy-loaded the message already. https://github.com/rails/rails/pull/24457

**Comment by jeremy** [comment]
(@mgodwin and @BenEddy are right on the money with the fix, too: Only touch the message within the mailer method itself. The `#mail` method is super handy for this.)

**Comment by jeremy** [comment]
The fix in #24457 isn't appropriate for a stable release line, so it'll be Rails 5 only. We could consider a deprecation warning for 4.x that allows the old behavior but squawks about it.
